File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/PowerSeriesRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.PowerRecord
import androidx.health.connect.client.records.PowerRecord.Sample
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.PowerMeasurementDto
import com.phamtunglam.health_connector_hc_android.pigeon.PowerSeriesRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [PowerRecord] object to a [PowerSeriesRecordDto].
 */
internal fun PowerRecord.toDto(): PowerSeriesRecordDto = PowerSeriesRecordDto(
    id = metadata.id,
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
    endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
    metadata = metadata.toDto(),
    samples = samples.map { sample ->
        PowerMeasurementDto(
            time = sample.time.toEpochMilli(),
            power = sample.power.toDto(),
        )
    },
)

/**
 * Converts a [PowerSeriesRecordDto] to a Health Connect [PowerRecord] object.
 */
internal fun PowerSeriesRecordDto.toHealthConnect(): PowerRecord = PowerRecord(
    samples = samples.map { sample ->
        Sample(
            time = Instant.ofEpochMilli(sample.time),
            power = sample.power.toHealthConnect(),
        )
    },
    startTime = Instant.ofEpochMilli(startTime),
    endTime = Instant.ofEpochMilli(endTime),
    startZoneOffset = startZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    endZoneOffset = endZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
    metadata = metadata.toHealthConnect(id),
)

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/power/power_series_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show HealthRecordId, PowerSeriesRecord, sinceV2_1_0;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/health_record_id_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/power/power_sample_mapper.dart';
import 'package:health_connector_hc_android/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show PowerSeriesRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [PowerSeriesRecord] to [PowerSeriesRecordDto].
@sinceV2_1_0
@internal
extension PowerSeriesRecordToDto on PowerSeriesRecord {
  /// Converts this power record to a DTO for platform transfer.
  PowerSeriesRecordDto toDto() {
    return PowerSeriesRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      metadata: metadata.toDto(),
      samples: samples.map((sample) => sample.toDto()).toList(),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
    );
  }
}

/// Converts [PowerSeriesRecordDto] to [PowerSeriesRecord].
@sinceV2_1_0
@internal
extension PowerSeriesRecordDtoToDomain on PowerSeriesRecordDto {
  /// Converts this DTO to a power record.
  PowerSeriesRecord toDomain() {
    return PowerSeriesRecord(
      id: id == null ? HealthRecordId.none : HealthRecordId(id!),
      startTime: DateTime.fromMillisecondsSinceEpoch(
        startTime,
        isUtc: true,
      ),
      endTime: DateTime.fromMillisecondsSinceEpoch(
        endTime,
        isUtc: true,
      ),
      metadata: metadata.toDomain(),
      samples: samples.map((dto) => dto.toDomain()).toList(),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
    );
  }
}
